02-peer-protocol.md

```diff
@@ -49,11 +68,502 @@ transaction is confirmed are also not persistent - until you know the script
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_input -->|       |
+        |   A   |<-(4)- tx_complete ----|   B   |
+        |       |--(5)- tx_add_output ->|       |
+        |       |<-(6)- tx_complete ----|       |
+        |       |--(7)- tx_complete --->|       |
+        +-------+                       +-------+
+
+#### *initiator* and *non-initiator*
+
+A, the *initiator*, contributes 2 inputs and an output that they
+then remove.  B, the *non-initiator*, contributes 1 input and an output,
+but waits until A adds a second input before contributing.
+
+Note that if A does not send a second input, the negotiation will end without
+B's contributions.
+
+        +-------+                         +-------+
+        |       |--(1)- tx_add_input ---->|       |
+        |       |<-(2)- tx_complete ------|       |
+        |       |--(3)- tx_add_output --->|       |
+        |       |<-(4)- tx_complete ------|       |
+        |       |--(5)- tx_add_input ---->|       |
+        |   A   |<-(6)- tx_add_input -----|   B   |
+        |       |--(7)- tx_remove_output >|       |
+        |       |<-(8)- tx_add_output ----|       |
+        |       |--(9)- tx_complete ----->|       |
+        |       |<-(10) tx_complete ------|       |
+        +-------+                         +-------+
+
+### The `tx_add_input` Message
+
+This message contains a transaction input.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u32`:`sequence`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input currently added to the
+    transaction
+  - MUST set `sequence` to be less than or equal to 4294967293 (`0xFFFFFFFD`)
+  - MUST NOT re-transmit inputs it has received from the peer
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received inputs to the transaction
+  - MUST fail the negotiation if:
+    - `sequence` is set to `0xFFFFFFFE` or `0xFFFFFFFF`
+    - the `prevtx` and `prevtx_vout` are identical to a previously added
+      (and not removed) input's
+    - `prevtx` is not a valid transaction
+    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
+    - the `scriptPubKey` of the `prevtx_vout` output of `prevtx` is not exactly a 1-byte push opcode (for the numeric values `0` to `16`) followed by a data push between 2 and 40 bytes
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - if has received 4096 `tx_add_input` messages during this negotiation
```

Comment:
Do you really expect that to happen? The way I see it, an `interactive-tx` session should be fast. Not necessarily in the dual funding case, but at least for splicing, we don't want the channel to be inactive for too long because of a slow `interactive-tx` session (because while the channel is inactive, you're not earning routing fees). In `eclair` we have an explicit 60 seconds timeout: if the `interactive-tx` construction isn't finished before that, we disconnect to discard it entirely.

I don't think it makes sense then to consider fee variations in such a small time interval, that would IMO add unnecessary complexity.

Reply:
> Do you really expect that to happen?

Yes. In a world where you have a competition between on-chain fees (based blockspace demand) and off-chain fees (based on channel topology) you will always have a price equilibrium seeked by economically rational actors.

Though I don’t see it happening in the coming years, maybe not even in the coming decade, so all good.